component/remote/vault/auth.go

```diff
@@ -0,0 +1,339 @@
+// (e.g., if renewal failed). vaultAuthenticate method may return a nil secret
+// if the authentication method does not generate a secret.
+type authMethod interface {
+	vaultAuthenticate(context.Context, *vault.Client) (*vault.Secret, error)
+}
+
+// AuthArguments defines a single authentication type in a remote.vault
+// component instance. These are embedded as an enum field so only one may be
+// set per AuthArguments.
+type AuthArguments struct {
+	AuthToken      *AuthToken      `river:"token,block,optional"`
+	AuthAppRole    *AuthAppRole    `river:"approle,block,optional"`
+	AuthAWS        *AuthAWS        `river:"aws,block,optional"`
+	AuthAzure      *AuthAzure      `river:"azure,block,optional"`
+	AuthGCP        *AuthGCP        `river:"gcp,block,optional"`
+	AuthKubernetes *AuthKubernetes `river:"kubernetes,block,optional"`
+	AuthLDAP       *AuthLDAP       `river:"ldap,block,optional"`
+	AuthUserPass   *AuthUserPass   `river:"userpass,block,optional"`
+	AuthCustom     *AuthCustom     `river:"custom,block,optional"`
+}
+
+func (a *AuthArguments) authMethod() authMethod {
+	switch {
+	case a.AuthToken != nil:
+		return a.AuthToken
+	case a.AuthAppRole != nil:
+		return a.AuthAppRole
+	case a.AuthAWS != nil:
+		return a.AuthAWS
+	case a.AuthAzure != nil:
+		return a.AuthAzure
+	case a.AuthGCP != nil:
+		return a.AuthGCP
+	case a.AuthKubernetes != nil:
+		return a.AuthKubernetes
+	case a.AuthLDAP != nil:
+		return a.AuthLDAP
+	case a.AuthUserPass != nil:
+		return a.AuthUserPass
+	case a.AuthCustom != nil:
+		return a.AuthCustom
+	}
+
+	panic("remote.vault: unreachable")
+}
+
+// AuthToken authenticates against Vault with a token.
+type AuthToken struct {
+	Token rivertypes.Secret `river:"token,attr"`
+}
+
+func (a *AuthToken) vaultAuthenticate(ctx context.Context, cli *vault.Client) (*vault.Secret, error) {
+	cli.SetToken(string(a.Token))
+	return nil, nil
+}
+
+// AuthAppRole authenticates against Vault with AppRole.
+type AuthAppRole struct {
+	RoleID        string            `river:"role_id,attr"`
+	Secret        rivertypes.Secret `river:"secret,attr"`
+	WrappingToken bool              `river:"wrapping_token,attr,optional"`
+	MountPath     string            `river:"mouth_path,attr,optional"`
+}
+
+func (a *AuthAppRole) vaultAuthenticate(ctx context.Context, cli *vault.Client) (*vault.Secret, error) {
+	secret := &approle.SecretID{FromString: string(a.Secret)}
+
+	var opts []approle.LoginOption
+	if a.WrappingToken {
+		opts = append(opts, approle.WithWrappingToken())
+	}
+	if a.MountPath != "" {
+		opts = append(opts, approle.WithMountPath(a.MountPath))
+	}
+
+	auth, err := approle.NewAppRoleAuth(a.RoleID, secret, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("auth.approle: %w", err)
+	}
+	s, err := cli.Auth().Login(ctx, auth)
+	if err != nil {
+		return nil, fmt.Errorf("auth.approle: %w", err)
+	}
+	return s, nil
+}
+
+// AuthAWS authenticates against Vault with AWS.
+type AuthAWS struct {
```

Comment:
I left it out on purpose, I don't think it makes sense to set a static nonce for the lifetime of the component, since the idea of a nonce is that it's used just once.